pallets/stableswap/src/tests/invariants.rs

```diff
@@ -0,0 +1,283 @@
+			)
+			.build()
+			.execute_with(|| {
+				let pool_id = get_pool_id_at(0);
+
+				let pool_account = AccountIdConstructor::from_assets(&PoolAssets(asset_a, asset_b), None);
+
+				assert_ok!(Stableswap::add_liquidity(
+					Origin::signed(BOB),
+					pool_id,
+					asset_a,
+					added_liquidity
+				));
+
+				let shares = Tokens::free_balance(pool_id.0, &BOB);
+
+				let amount_withdrawn = withdraw_percentage.mul_floor(shares);
+
+				let asset_a_reserve = Tokens::free_balance(asset_a, &pool_account);
+				let asset_b_reserve = Tokens::free_balance(asset_b, &pool_account);
+				assert_ok!(Stableswap::remove_liquidity(Origin::signed(BOB), pool_id, amount_withdrawn));
+
+				let new_asset_a_reserve = Tokens::free_balance(asset_a, &pool_account);
+				let new_asset_b_reserve = Tokens::free_balance(asset_b, &pool_account);
+
+				assert_eq_approx!(FixedU128::from((asset_a_reserve,asset_b_reserve)),
+					FixedU128::from((new_asset_a_reserve,new_asset_b_reserve)),
+					FixedU128::from_float(0.0000000001),
+					"Price has changed after remove liquidity");
+			});
+	}
+}
+
+proptest! {
+	#![proptest_config(ProptestConfig::with_cases(1000))]
+	#[test]
+	fn sell_invariants(
+		initial_liquidity in asset_reserve(),
+		amount in trade_amount(),
+		amplification in amplification(),
+	) {
+		let asset_a: AssetId = 1000;
+		let asset_b: AssetId = 2000;
+
+		ExtBuilder::default()
+			.with_endowed_accounts(vec![
+				(BOB, asset_a, amount),
+				(ALICE, asset_a, initial_liquidity),
+				(ALICE, asset_b, initial_liquidity),
+			])
+			.with_registered_asset("one".as_bytes().to_vec(), asset_a)
+			.with_registered_asset("two".as_bytes().to_vec(), asset_b)
+			.with_pool(
+				ALICE,
+				PoolInfo::<AssetId> {
+					assets: PoolAssets::new(asset_a,asset_b),
+					amplification,
+					fee: Permill::from_percent(0),
+				},
+				InitialLiquidity{ account: ALICE, asset:asset_a, amount:initial_liquidity},
+			)
+			.build()
+			.execute_with(|| {
+				let pool_id = get_pool_id_at(0);
+
+				let pool_account = AccountIdConstructor::from_assets(&PoolAssets(asset_a, asset_b), None);
+
+				let asset_a_reserve = Tokens::free_balance(asset_a, &pool_account);
+				let asset_b_reserve = Tokens::free_balance(asset_b, &pool_account);
+				let ann: Balance = amplification as u128 * 2 * 2;
+				let d_prev = calculate_d::<128u8>(&[asset_a_reserve,asset_b_reserve], ann, 1u128).unwrap();
+
+				assert_ok!(Stableswap::sell(
+					Origin::signed(BOB),
+					pool_id,
+					asset_a,
+					asset_b,
+					amount,
+					0u128, // not interested in this
+				));
+
+				let asset_a_reserve = Tokens::free_balance(asset_a, &pool_account);
+				let asset_b_reserve = Tokens::free_balance(asset_b, &pool_account);
+				let ann: Balance = amplification as u128 * 2 * 2;
+				let d = calculate_d::<128u8>(&[asset_a_reserve,asset_b_reserve], ann, 1u128).unwrap();
+
+				assert!(d >= d_prev);
+				assert!(d - d_prev <= 10u128);
```

Comment:
where does the 10 come from? seems pretty magic to me :-D

Reply:
Magic number discussed with @poliwop .  It is maximum allowed difference between the D before and after a trade is done.

Agreed that it would be better to create a tolerance constant with a comment.